torch/csrc/distributed/c10d/init.cpp

```diff
@@ -1018,6 +1051,16 @@ that adds a prefix to each key inserted to the store.
               [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
                  const std::vector<at::Tensor>& tensors,
                  const ::c10d::AllreduceOptions& opts) {
+                if (opts.reduceOp.op_ == ::c10d::ReduceOp::PREMUL_SUM) {
+#if defined(ENABLE_NCCL_PREMUL_SUM_SUPPORT)
```

Comment:
We should have this sort of hacks in the backend agnostic part of the code.
It should be up to each backend to detect and fail if the user tries an unsupported reducion

Reply:
This is the only way I could find to deal with the requirement of `::c10d::ops:allreduce`, assuming `struct ReduceOp` having a scaling factor is a correct way to go.

Reply:
@kumpera Do you mean we should *not* have this sort of hacks in the backend agnostic part of the code?

Reply:
I agree that once the dispatcher->implementation route is in place, this kind of check can be put into the CUDA/NCCL specific implementation, rather than for the general ProcessGroup part.